tests/integration/ha_tests/test_storage.py

```diff
@@ -0,0 +1,343 @@
+    and then we deploy a new application with storage reuse and check that the
+    storage has been reused.
+    """
+    app_name = await get_app_name(ops_test)
+    if storage_type(ops_test, app_name) == "rootfs":
+        pytest.skip(
+            "reuse of storage can only be used on deployments with persistent storage not on rootfs deployments"
+        )
+
+    writes_results = await stop_continous_writes(ops_test, app_name=app_name)
+    time.sleep(70)  # Leave some time to write the oplog + lock (60s + a few more)
+    unit_ids = [unit.name for unit in ops_test.model.applications[app_name].units]
+    storage_ids = {}
+
+    remaining_units = len(unit_ids)
+    for unit_id in unit_ids:
+        storage_ids[unit_id] = storage_id(ops_test, unit_id)
+        await ops_test.model.applications[app_name].destroy_unit(unit_id)
+        # Give some time to remove the unit. We don't use asyncio.sleep here to
+        # leave time for each unit to be removed before removing the next one.
+        # time.sleep(60)
+        remaining_units -= 1
+        await ops_test.model.wait_for_idle(
+            apps=[app_name],
+            status="active",
+            timeout=TIMEOUT,
+            idle_period=20,
+            wait_for_exact_units=remaining_units,
+        )
+
+    # Wait until all apps are cleaned up
+    await ops_test.model.wait_for_idle(apps=[app_name], timeout=TIMEOUT, wait_for_exact_units=0)
+
+    await ops_test.model.remove_application(app_name, block_until_done=True, destroy_storage=False)
+
+    new_app_name = OTHER_APP_NAME
+
+    await ops_test.model.deploy(
+        database_charm,
+        application_name=new_app_name,
+        num_units=1,
+        attach_storage=[tag.storage(storage_ids[unit_ids[0]])],
+    )
+    deployed_units = 1
+    await ops_test.model.wait_for_idle(
+        apps=[new_app_name],
+        status="active",
+        timeout=TIMEOUT,
+        idle_period=60,
+        wait_for_exact_units=deployed_units,
+    )
+
+    for unit_id in unit_ids[1:]:
+        deployed_units += 1
+        await ops_test.model.applications[new_app_name].add_unit(
+            count=1, attach_storage=[tag.storage(storage_ids[unit_id])]
+        )
+        await ops_test.model.wait_for_idle(
+            apps=[new_app_name],
+            status="active",
+            timeout=TIMEOUT,
+            idle_period=60,
+            wait_for_exact_units=deployed_units,
+        )
+
+    assert len(ops_test.model.applications[new_app_name].units) == len(storage_ids)
+
+    # check if previous volumes are attached to the units of the new cluster
+    new_storage_ids = []
+    for unit in ops_test.model.applications[new_app_name].units:
+        new_storage_ids.append(storage_id(ops_test, unit.name))
+
+    assert sorted(storage_ids.values()) == sorted(new_storage_ids), "Storage IDs mismatch"
+
+    actual_writes = await count_writes(ops_test, app_name=new_app_name)
+    assert writes_results["number"] == actual_writes
```

Comment:
same checks as suggested above

Reply:
I don't think you have suggested a check above, maybe a comment got lost?